File: duckdb-odbc-python-demo/app/duckdb_conn.py
"""Single shared DuckDB connection, the Python analogue of DuckDbConfig.

Why one connection? The ODBC extension stores its connection handle in a DuckDB
*session variable* (SET VARIABLE conn = odbc_connect(...)), which is scoped to a
single DuckDB connection. In the Java demo this forced a SingleConnectionDataSource
instead of a pool. In Python it is simpler: a `duckdb.connect()` object *is* one
connection, so we create it once and share it. It also means each new connection
would be its own in-memory database, so sharing the handle is essential.

DuckDB's Python connection is not thread-safe for concurrent use, so a lock
serializes access - fine for a demo, and honest about the single-connection model.
"""
from threading import Lock
from typing import Any

import duckdb

from .config import settings

_conn: duckdb.DuckDBPyConnection = duckdb.connect(settings.duckdb_database)
_lock = Lock()


def query(sql: str, params: list[Any] | None = None) -> list[dict[str, Any]]:
    """Run a query and return rows as a list of dicts (like JdbcTemplate.queryForList)."""
    with _lock:
        cur = _conn.execute(sql, params) if params else _conn.execute(sql)
        columns = [d[0] for d in cur.description]
        return [dict(zip(columns, row)) for row in cur.fetchall()]


def execute(sql: str, params: list[Any] | None = None) -> None:
    """Run a statement that returns no rows (like JdbcTemplate.execute/update)."""
    with _lock:
        _conn.execute(sql, params) if params else _conn.execute(sql)

File: odbc-scenarios/duckdb-odbc-node-demo/src/duckdbConn.js
// Single shared DuckDB connection, the Node analogue of DuckDbConfig.
//
// Why one connection? The ODBC extension stores its connection handle in a
// DuckDB *session variable* (SET VARIABLE conn = odbc_connect(...)), which is
// scoped to a single DuckDB connection. In the Java demo this forced a
// SingleConnectionDataSource instead of a pool. Here it is simpler: a
// DuckDBConnection object *is* one connection, so we create it once and share
// it. Each new connection to ':memory:' would be its own database, so sharing
// the handle is essential.
//
// A promise queue serializes statements: DuckDB runs them on one connection and
// SET VARIABLE state must not interleave across concurrent requests. This keeps
// the single-connection model honest, matching the Java and Python demos.

import { DuckDBInstance } from '@duckdb/node-api';
import { config } from './config.js';

let connection;
let tail = Promise.resolve(); // serializes access to the single connection

export async function connect() {
  const instance = await DuckDBInstance.create(config.duckdb.database);
  connection = await instance.connect();
}

// Serialize `fn` behind any in-flight work on the shared connection.
function enqueue(fn) {
  const run = tail.then(fn, fn);
  // Keep the chain alive even if a statement rejects.
  tail = run.then(
    () => undefined,
    () => undefined
  );
  return run;
}

// Run a query and return rows as plain JSON-safe objects (like
// JdbcTemplate.queryForList). getRowObjectsJson handles DuckDB types that
// aren't natively JSON-serializable: BIGINT -> string, DECIMAL -> string.
export function query(sql, params) {
  return enqueue(async () => {
    if (params && params.length) {
      const prepared = await connection.prepare(sql);
      bindAll(prepared, params);
      const reader = await prepared.runAndReadAll();
      return reader.getRowObjectsJson();
    }
    const reader = await connection.runAndReadAll(sql);
    return reader.getRowObjectsJson();
  });
}

// Run a statement that returns no rows we care about (like
// JdbcTemplate.execute/update).
export function execute(sql, params) {
  return enqueue(async () => {
    if (params && params.length) {
      const prepared = await connection.prepare(sql);
      bindAll(prepared, params);
      await prepared.run();
      return;
    }
    await connection.run(sql);
  });
}

// Bind positional parameters. DuckDB's Node API is typed per-bind; the demo
// only ever binds strings (ODBC connection string, credentials, SQL text).
function bindAll(prepared, params) {
  params.forEach((value, i) => prepared.bindVarchar(i + 1, String(value)));
}
